rust-symcrypt/src/ecdh.rs

```diff
@@ -0,0 +1,352 @@
+/* EcDh functions. For further documentation please refer to symcrypt.h */
+
+use crate::curve_type::*;
+use crate::errors::SymCryptError;
+use std::vec;
+use symcrypt_sys;
+
+pub struct EcDh {
+    curve_type: CurveType,
+    expanded_curve: Option<symcrypt_sys::PSYMCRYPT_ECURVE>,
+    key: Option<symcrypt_sys::PSYMCRYPT_ECKEY>,
+}
+
+impl EcDh {
+    /// EcDh::new() returns a EcDh struct that has a private/public key pair.
+    pub fn new(curve: CurveType) -> Result<Self, SymCryptError> {
+        let mut instance = EcDh {
+            curve_type: curve,
+            expanded_curve: None,
+            key: None,
+        };
+
+        unsafe {
+            // SAFETY: FFI calls
+            let curve_ptr =
+                symcrypt_sys::SymCryptEcurveAllocate(convert_curve(instance.curve_type), 0);
+            if curve_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            instance.expanded_curve = Some(curve_ptr);
+
+            let key_ptr = symcrypt_sys::SymCryptEckeyAllocate(curve_ptr);
+            if key_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            instance.key = Some(key_ptr);
+
+            symcrypt_sys::SymCryptEckeySetRandom(
+                symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH,
+                instance.key.unwrap(),
+            );
+        }
+        Ok(instance)
+    }
+
+    /// ExDh::from_public_key_bytes function returns a EcDh struct that only has a public key attached.
+    pub fn from_public_key_bytes(
+        curve: CurveType,
+        public_key: &[u8],
+    ) -> Result<Self, SymCryptError> {
+        let mut instance = EcDh {
+            curve_type: curve,
+            expanded_curve: None,
+            key: None,
+        };
+
+        let num_format = if instance.curve_type == CurveType::Curve25519 {
+            symcrypt_sys::_SYMCRYPT_NUMBER_FORMAT_SYMCRYPT_NUMBER_FORMAT_LSB_FIRST
+        } else {
+            symcrypt_sys::_SYMCRYPT_NUMBER_FORMAT_SYMCRYPT_NUMBER_FORMAT_MSB_FIRST
+        };
+
+        let ec_point_format = symcrypt_sys::_SYMCRYPT_ECPOINT_FORMAT_SYMCRYPT_ECPOINT_FORMAT_XY;
+
+        unsafe {
+            // SAFETY: FFI calls
+            let curve_ptr =
+                symcrypt_sys::SymCryptEcurveAllocate(convert_curve(instance.curve_type), 0);
+            if curve_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            instance.expanded_curve = Some(curve_ptr);
+
+            let key_ptr = symcrypt_sys::SymCryptEckeyAllocate(instance.expanded_curve.unwrap());
+            if key_ptr.is_null() {
+                return Err(SymCryptError::AuthenticationFailure);
+            }
+
+            match symcrypt_sys::SymCryptEckeySetValue(
+                std::ptr::null_mut(),
+                0,
+                public_key.as_ptr(),
```

Comment:
>public_key.as_ptr() [](http://example.com/codeflow?start=16&length=19)

Just curious. What is as_ptr() for an empty slice?

Shouldn't be a problem, where .len() will be 0 so it shouldn't be used?